Support Vite 3 and latest babel-plugin-relay

tobias-tengler opened the pull request:
I tried to use this plugin in a new Vite app and noticed that it doesn't work, due to the usage of `require.resolve`. While fixing this I also noticed that it has a dependency on a fixed version of `babel-plugin-relay`, which is kind of cumbersome to maintain, so I:

- replaced `require.resolve` with just the plugin name
- made `babel-plugin-relay` a peer dependency
- forwarded the filename to `transformSync` in order to support `artifactDirectory` imports
- updated all of the packages to their latest versions
- changed the package type to "module" and used tsup to generate both commonjs and esm output
- updated the README with more detailed instructions
- added example project for Vite 3
- switched to package linking instead of `file:..`
- fixed an issue in the test pipeline

Closes #309 
Closes #211

tobias-tengler commented on tsconfig.json:
```diff
@@ -10,6 +10,7 @@
     "sourceMap": true,
-    "target": "es2018"
+    "target": "es2018",
+    "skipLibCheck": true
```
This speeds up the build a bunch, because it only validates our type definitions and not the type definitions in referenced dependencies. This settings is recommended by the TypeScript team.

tobias-tengler commented on src/plugin.ts:
```diff
@@ -5,20 +5,25 @@ export default {
           ],
         ],
         code: true,
+        filename: id,
```
This forwards the filename to `transformSync`, so `babel-plugin-relay` can correctly compute import paths, when using `artifactDirectory`.

tobias-tengler commented:
@oscartbeaumont If it's alright for you, I would also credit myself in the README :)

oscartbeaumont commented:
Feel free to!

tobias-tengler commented:
Whoops. Forgot to update the tests. I'm on it!

tobias-tengler commented:
@oscartbeaumont Okay tests should be fixed :)

tobias-tengler commented:
@oscartbeaumont Sorry for the pings, but tests need to be run again :D I think the playwright installation issue should now also be fixed

tobias-tengler commented:
@oscartbeaumont The tests should now succeed!

I think it should be fine to go ahead and merge / release it. 
It should be published as a major version bump, as it will break existing projects, if they are not on Vite 3 or do not have `babel-plugin-relay` and `graphql` installed in their project.

The migration path for users of v1 of this library is:
1. Update project to Vite 3: https://vitejs.dev/guide/migration.html
1. `yarn install vite-plugin-relay@latest -D`
2. `yarn install babel-plugin-relay@latest graphql -D` (`babel-plugin-relay` version should match the version of the installed Relay packages, but must be >= 13.0.1 - it's probably best to just upgrade all Relay packages to the latest versoin)

New users can also use my [create-relay-app](https://github.com/tobias-tengler/create-relay-app) project to correctly setup Relay with this plugin in their projects.

tobias-tengler commented:
@oscartbeaumont I've re-added the Vite 2 example and changed the code to support both Vite 2 and 3. The pipeline should now run tests against both versions.

Users migrating now do not need to update their version.

Could you review the code and give me feedback whether we can merge it or not? Would love to get this in asap :)

oscartbeaumont commented:
Just give me till the end of this week to review. I wanna give it a proper test given so much has changed but I have been super busy.

oscartbeaumont commented:
Thanks for all of your work by the way! I have been meaning to give this project some love but just haven't found the time for it.

tobias-tengler commented:
Okay great :)
I've just switched to package linking instead of `file:./../..`, since it causes the local cache to fill up quickly and took upwards of 5min sometimes for me to install the newest version.
So now when developing run `yarn link` in the root directory and then `yarn link vite-plugin-relay` in the example directories and they should always point to the current version of the code. I also updated the workflow to do the same.

tobias-tengler commented:
@oscartbeaumont Just wanted to remind you to review this, since it's the end of the week.

tobias-tengler commented:
@oscartbeaumont bump

tobias-tengler commented:
@oscartbeaumont Do you think you'll get a chance to review it this weekend?

tobias-tengler commented:
@oscartbeaumont Sorry for being annoying, but it would be really great to get this in asap.
Would it help you if I broke down the changes into smaller PRs?

oscartbeaumont commented:
I am really sorry for taking so long to review this. I know how annoying it is. Between my full-time job and being burnt out I just didn't get around to it until now.

I moved to a pnpm workspace to make installation quicker and not require any linking.

I have added @tobias-tengler as a GitHub contributor. When you create a GitHub release it will automatically push to npm. Would be nice to still use GH Issues/PRs to track changes before releases though. Also feel free to add me on Discord (oscartbeaumont#0004) if you wanna chat.
